### ISSUE TITLE
[Issue]: Connector Health Check does not Work in Encrypted Mode

### ISSUE THREAD

**Issue by RussGarratt** [paypal/mirakl-hyperwallet-connector#117]
### Describe the Issue

We struggled to deploy the connector to our live environment due to the health check URL failing.

There seem to be a number of issues:

_DefaultHyperwalletSDKUserService:_

This class is called by the health check passing the "getRootProgramToken()"

	@Override
	public HyperwalletProgram getRootProgram() {
		return **getHyperwalletInstanceByHyperwalletProgram**(userHyperwalletApiConfig.getRootProgramToken())
				.getProgram(userHyperwalletApiConfig.getRootProgramToken());
	}

(The below is a Map of hyperwalletProgram to token, e.g. DEFAULT -> prg1234567)
However, rootToken is as we see below the prg number, so this seems to be null always as it is trying to get the key of the map as the prg number and not 'DEFAULT'
It looks like the other method should have been called, (getHyperwalletInstanceByProgramToken) rather than byHyperWalletProgram

	@Override
	public Hyperwallet **getHyperwalletInstanceByHyperwalletProgram**(final String hyperwalletProgram) {
		//@formatter:off
		final String programUserToken = Optional.ofNullable(userHyperwalletApiConfig.getTokens())
				.map(tokens -> tokens.get(hyperwalletProgram))
				.orElse(null);
		//@formatter:on
		return new Hyperwallet(this.userHyperwalletApiConfig.getUsername(), this.userHyperwalletApiConfig.getPassword(),
				programUserToken, this.userHyperwalletApiConfig.getServer(), encryption);
	}

UserHyperwalletApiConfig:
	@Value("#{'${infrastructure.hyperwallet.api.hyperwalletprogram.**rootToken**}'}")
	private String rootProgramToken;

Infrastructure.properties:

infrastructure.hyperwallet.api.hyperwalletprogram.rootToken     =  
${PAYPAL_HYPERWALLET_PROGRAM_TOKEN_USERS_DEFAULT}    (The prg number e.g. prg1245678765432)

Even if I fixed this so that we are returning what appears to be the correct token, the actual Health Check call itself fails due to us having the encrypted profile.

Once token is retrieved, the get request fails,
This seems to be because:

Hyperwallet.java

    /**
     * Get Program
     *
     * @param programToken Program token
     * @return HyperwalletProgram
     */
    public HyperwalletProgram getProgram(String programToken) {
        if (StringUtils.isEmpty(programToken)) {
            throw new HyperwalletException("Program token is required");
        }
        return apiClient.get(url + "/programs/" + programToken, HyperwalletProgram.class);
    }

calls the apiClient.get,
Which calls buildGetRequest:

 private Request buildGetRequest(final String url) {
        Request request = new Request(url, connectionTimeout, readTimeout, proxy, proxyUsername, proxyPassword);
        return populateCommonHTTPHeader(request)
                .addHeader(ACCEPT, getContentType());
    }

    private String getContentType() {
        return isEncrypted ? **APPLICATION_JOSE_JSON** : APPLICATION_JSON;
    }

Which then complains that there is:

 No JWK found at the configured URL for this url.

If the get Requests weren't to check the content type (I guess as there  is no encrypted body necessary) - that would probably work? 

### Environment

Live/Production

### Version

4.18.1

### Expected Behavior

Able to start server - health checks pass

### Actual Behavior

Kubernetes marks pods as unavailable as health check fails

### Steps to Reproduce

As above, or perform a simple check to the Hyperwallet API and in Postman set the content type as application/jose+json

e.g:

https://uat-api.paylution.com/rest/v4/programs/prg-<prgnumber>

In postman, set basic auth credentials and set content type.

{
  "errors" : [ {
    "message" : " No JWK found at the configured URL for this url.",
    "code" : "JWT_ERROR"
  } ]
}

### Pre-conditions

As above

### Relevant log output

Not needed

I would have submitted a pull-request but I am not sure what the intention is here due to all the work to support multiple programs!

Best Regards,

Russ

**Comment by max6001** [comment]
Hi @RussGarratt thanks for reaching out.
We're looking into it and will get back to you asap.

Best,

**Comment by javierpoloe2y** [comment]
Hi @RussGarratt,

You are right, we have a bug and we should have called `getHyperwalletInstanceByProgramToken`. Fortunately, this doesn't impact `getRootProgram` since Hyperwallet SDK allows to create a new instance of `Hyperwallet` with a null token and as you see in:

```java
public HyperwalletProgram getProgram(String programToken) {
    if (StringUtils.isEmpty(programToken)) {
        throw new HyperwalletException("Program token is required");
    }
    return apiClient.get(url + "/programs/" + programToken, HyperwalletProgram.class);
}
```

is retrieving the program token passed as a parameter, which in this case is `userHyperwalletApiConfig.getRootProgramToken()` which as you said has the same `prg-*` value as the DEFAULT program token.

The error:

```json
{
  "errors" : [ {
    "message" : " No JWK found at the configured URL for this URL.",
    "code" : "JWT_ERROR"
  } ]
}
```

is returned from Hyperwallet because it can't recover your JWK public keys, which are exposed at the `/jwkset` endpoint (you can take a look at `JwkSetController`).

Please check that your `/jwkset` endpoint is configured in Hyperwallet as the JWK URI and that the endpoint is accessible from Hyperwallet servers.

**Comment by RussGarratt** [comment]
I don't believe you are correct sorry @javierpoloe2y  regarding the JWK stuff

Please take a look at your switch for getContentType:

private String getContentType() {
    return isEncrypted ? **APPLICATION_JOSE_JSON** : APPLICATION_JSON;
}

Hyperwallet.java

and perform a GET Request to the program:

![Issue With Content Type](https://user-images.githubusercontent.com/104622039/224305990-3722106f-cee2-4041-aa0d-516dfe3b1a95.png)


![Correct](https://user-images.githubusercontent.com/104622039/224306309-15d7da2c-44b9-4591-882c-dbf5496f109f.png)

**Comment by javierpoloe2y** [comment]
Hi @RussGarratt ,

Since you have enabled the encryption in the connector, it's making the requests using the `application/jose+json` content type (this is handled by the Hyperwallet SDK used by the connector where the `getContentType` code snippet comes from).

So as you see in your test in Postman when you make the request using `application/jose+json` it gives you the same error that you receive in the connector. If you use the content type `application/json` the Hyperwallet server understands that you don't want to use encryption and it doesn't fail. So it looks that something is misconfigured related to the encryption.

You can read more about this in [the official Hyperwallet API encryption documentation](https://docs.hyperwallet.com/content/api/v4/overview/payload-encryption#rest-api-authentication). The connector project README also has a section that explains how to configure the encryption (also available [here])(https://paypal.github.io/mirakl-hyperwallet-connector/component-hmc/4.18.1/readme/readme.html#_payload_encryption)

So please verify that you have set up the encryption correctly in Hyperwallet, perhaps the Hyperwallet team that help you with the onboarding can help you with this also.

**Comment by RussGarratt** [comment]
Hi @javierpoloe2y 

I believe we have the encryption profile setup correctly and have been using it up until now on previous versions.
I still think there is a bug in perhaps the Hyperwallet SDK?

Can you please try this and see whether it works for you?

This is essentially what your health checker is doing, so I created a small main class just to test:

`	public static void main(final String[] args) throws UnknownHostException {

		boolean fileExists = Files.exists(Paths.get("private-jwk-key-location"));
		System.out.println("Does the File exist? " + fileExists);
		System.out.println("Setting up encryption");

		HyperwalletEncryption encryption = new HyperwalletEncryption(
				JWEAlgorithm.RSA_OAEP_256,
				com.nimbusds.jose.JWSAlgorithm.RS256,
				EncryptionMethod.A256CBC_HS512,
				"<private-jwk-key-location>",
				"https://uat-api.paylution.com/jwkset",
				5);

		System.out.println("Creating client with encryption");

		Hyperwallet hwClient = new Hyperwallet("<username>", "<password>",
				"<prg-token(USERS_DEFAULT)>", "https://uat-api.paylution.com", encryption);

		System.out.println("Requesting Program");

		hwClient.getProgram("<prg-token(USERS_DEFAULT)>");
	}`

What I get is:

> Task :Main.main() FAILED
Does the File exist? true
Setting up encryption
Creating client with encryption
Requesting Program
Exception in thread "main" com.hyperwallet.clientsdk.HyperwalletException:  No JWK found at the configured URL for this url.
	at com.hyperwallet.clientsdk.util.HyperwalletApiClient.checkErrorResponse(HyperwalletApiClient.java:325)
	at com.hyperwallet.clientsdk.util.HyperwalletApiClient.processResponse(HyperwalletApiClient.java:281)
	at com.hyperwallet.clientsdk.util.HyperwalletApiClient.get(HyperwalletApiClient.java:82)
	at com.hyperwallet.clientsdk.Hyperwallet.getProgram(Hyperwallet.java:2387)
	at Main.main(Main.java:36)

**Comment by RussGarratt** [comment]
Since that is on the response side, do Hyperwallet need to configure URL / encryption mappings?
I thought we simply needed them to enable the encryption profile / set the URL to our public key?
( I made sure my 'local' keys were the same as our test env that we have configured up against HW)

**Comment by javierpoloe2y** [comment]
Hi @RussGarratt,

This is an extract from [the Hyperwallet Payload Encryption documentation](https://docs.hyperwallet.com/content/api/v4/overview/payload-encryption#rest-api-authentication) that I have referred to in a previous post:

> Clients choosing JOSE may submit requests from any IP address, However, in addition to HTTP BASIC authentication and TLS, request payloads must be signed and encrypted. Clients will also be required to decrypt then validate the signature of all responses to access the payload. **Clients must host their public keys as a JWK key set and provide the JWK keyset URI while onboarding.** Clients should use different keys for UAT (test) and PROD. The keys should be hosted at different URIs. Client public keys must be hosted with TLS with a valid certificate.

As you see the URI for your public JWK keys must be configured in Hyperwallet and must be accessible from Hyperwallet servers. Please reach out Hyperwallet support so they can help you with that. Also as I explained before, the connector exposes an `/jwkset` endpoint with the public keys (for example `http://localhost:8080/jwkset`)

**Comment by RussGarratt** [comment]
Hi @javierpoloe2y 

Would like to say thanks for re-stressing the point about the jwkset setup!

I checked the /jwkset  on our UAT environment and found it was 404'ing!

So here is what happened:

We deployed 4.18.1 (vanilla) to our 'UAT' environment (which has the configured jwkset URL to Hyperwallet)
Due to the issues upgrading to 4.18.1 it seemed that the initial Health check failed (program token)
This meant that the pod was actually deployed (logging jobs etc) but that the load balancer had not marked the pod as active, meaning the RESTful endpoints were not accessible to Hyperwallet.

The tests from my local machine were being sent to HW but HW was unable to access the /jwkset URL on our configured UAT endpoint resulting in the error message we saw.

Resolved by:
Deploying out an earlier version of the connector
Tested locally -> All worked Ok after /jwkset accessible (requires key share / same keys in local as UAT environment)
Submitted changes to retrieve the 'correct' DEFAULT token ( as mentioned above) -> Not sure if this was necessary, but seems so.
Deployed out 4.18.1 (plus changes above) to UAT environment -> All OK!

Will check the instructions for intelliJ setup, so thanks for that.
We have been building a local docker image from the gradle tasks and running as a docker image supplying ENV vars to docker.

So we are now running 4.18.1 with:

* Removal of :

	@ConditionalOnMissingBean
	@Bean
	public BuildProperties buildProperties() {
		return new BuildProperties(new Properties());
	}
From WebConnectorApplication

And: Changing DefaultHyperwalletSDKUserService to:


	@Override
	public Hyperwallet getHyperwalletInstanceByHyperwalletProgram(final String hyperwalletProgram) {
		return new Hyperwallet(this.userHyperwalletApiConfig.getUsername(), this.userHyperwalletApiConfig.getPassword(),
				hyperwalletProgram, this.userHyperwalletApiConfig.getServer(), encryption);
	}

Best regards,
Russ
